apps/emqx_rule_engine/src/emqx_rule_actions.erl

```diff
@@ -232,3 +243,70 @@ format_pub_props(UserPropertiesTks, Selected, Env) ->
                 replace_simple_var(UserPropertiesTks, Selected, #{})
         end,
     #{'User-Property' => UserProperties}.
+
+format_mqtt_properties(MQTTPropertiesTemplate, Selected, Env) ->
+    #{metadata := #{rule_id := RuleId}} = Env,
+    MQTTProperties0 =
+        maps:fold(
+            fun(K, Template, Acc) ->
+                try
+                    V = emqx_placeholder:proc_tmpl(Template, Selected),
+                    Acc#{K => V}
+                catch
+                    Kind:Error ->
+                        ?SLOG(
+                            error,
+                            #{
+                                msg => "bad_mqtt_prop_value",
+                                rule_id => RuleId,
+                                reason => {Kind, Error},
+                                property => K,
+                                selected => Selected
+                            }
+                        ),
+                        Acc
+                end
+            end,
+            #{},
+            MQTTPropertiesTemplate
+        ),
+    coerce_properties_values(MQTTProperties0, Env).
+
+ensure_int(B) when is_binary(B) -> binary_to_integer(B);
+ensure_int(I) when is_integer(I) -> I.
```

Comment:
better error handling by raising a `throw` exception here
then `catch throw:Reason` can produce a more readable error and less stacktrace in logs.

Reply:
sorry, stacktrace is not logged.
but for exceptions other than `throw`, we should probably log stacktrace?